### DISCUSSION TITLE
Adding Ribbon Geometry in tri_mesh ?

### DISCUSSION THREAD

**Discussion by jbousquie** [asny/three-d-asset#46] [Ideas]
Hi,
I just achieved to code a prototype of a parametric shape, something simple called "ribbon", that allows to build a mesh from a vector of paths, each path being a vector of successive vec3 along a line.
![ribbon](https://github.com/user-attachments/assets/89801f70-294e-4071-b582-97dfd16d7c56)

Q&D prototype : https://github.com/jbousquie/threeribbon

Would you accept that I add this kind of geometry in this file : https://github.com/asny/three-d-asset/blob/main/src/geometry/tri_mesh.rs  as the provided build-in geometries aren't that numerous ?
The ribbon could be a versatile tool to build then algorithmatically other shapes like tubes, extrusions, lathes, etc.

Here are example-like to what I intend to do (in a simpler way with just one input parameter, the path_array) : 
https://doc.babylonjs.com/features/featuresDeepDive/mesh/creation/param/ribbon/
https://doc.babylonjs.com/features/featuresDeepDive/mesh/creation/param/ribbon_maths/#example-pg--F6JW5W-18

and it will be morphable too : 

https://playground.babylonjs.com/#1MSEBT#3

https://playground.babylonjs.com/#ACKC2#1

Please just let me know if you're interested 
regards

-- 
Jerome

**Comment by asny** [comment]
It is actually very much on purpose that I didn't add a lot of different geometry generation methods to this library, it's not really the purpose of this crate. The purpose is to be able to serialize/deserialize 3D content, so I would expect people to use it for that in 99% of cases. It's easy to generate some geometry outside of the crate and then construct a `TriMesh` manually from the result, so that is what I would suggest. I added a few simple geometry generation methods just to make it easy to get started, I didn't plan to add more. Maybe you can consider adding the functionality to [tri-mesh](https://github.com/asny/tri-mesh) or some other geometry processing library?

**Reply by jbousquie** [reply]
ak ok
If it was to be added in tri-mesh, where (what file/module) would it be in this lib ?
I haven't read all the code so far, but I just couldn't find a location where some meshes are generated in tri-mesh... that's why I assumed it whoud be among the other mesh builders in tree-d-assets :-)

**Reply by asny** [reply]
That's a really good point, I guess I didn't add any geometry generation functionality to `tri-mesh` yet. You can just add it as functionality directly to the `Mesh` class in `mesh.rs`. However, the problem might be that the `tri-mesh` crate doesn't support uv coordinates because it significantly complicates the mesh operations, I guess you'll need those? I'm actually not really sure were to put this, I didn't really want the `three-d-asset` crate to be about generating content, just about loading/saving assets. Maybe there's another library out there for generating meshes and where the result can be converted into a `TriMesh`?

**Reply by jbousquie** [reply]
Ok, no problem.
As the ribbon is for now just a feature for my needs, I'll put it somewhere else (by the time I learn how to make a crate) either in a mesh builder lib somewhere or in my own one, dedicated to build TriMeshes algorithmatically

**Reply by asny** [reply]
Sounds like a good plan. It feels like it belongs in another crate and I don't have the time or energy to start a new one 😄
